File: Kth Largest Element in an Array.js
var findKthLargest = function(nums, k) {
    let pivot, pivotIdx;
    let finalIdx = nums.length -k
    let start = 0;
    let end = nums.length-1;
    while(start <= end){
        pivot = Math.floor(Math.random()*(end-start+1)) +start;
        pivotIdx = pivotHelper(pivot, start, end);
        
        if(pivotIdx === finalIdx) return nums[pivotIdx];
        if(pivotIdx > finalIdx) end = pivotIdx-1;
        else start = pivotIdx+1;
    }
    
    function pivotHelper(pivot, start, end){
        let pivotVal = nums[pivot];
        swap(pivot, end);
        let i = 0;
        for(let j = 0; j < end; j++){
            if(nums[j] <= pivotVal){
                swap(i,j);
                i++;
            }
        }
        swap(i, end);
        return i;
    }
    
    function swap(i, j){
        let temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp
    }
};

File: Kth Largest Element in an Array.py
class Solution:
    def quickSort(self,nums, k, left, right):
            pivot = random.randint(left, right)
            # swap the pivot element with the left most element
            nums[pivot], nums[left] = nums[left], nums[pivot]
            write = left+1
            for read in range(left+1, right+1):
                if nums[read] <= nums[left]:
                    nums[read], nums[write] = nums[write], nums[read]
                    write += 1
            
            nums[left], nums[write-1] = nums[write-1], nums[left]

            if len(nums) - (write - 1) == k:
                return nums[write - 1]
            if len(nums) -(write - 1)  < k:  
                return self.quickSort(nums, k, left, write-2)
            else:
                return self.quickSort(nums, k, write, right)

    def findKthLargest(self, nums: List[int], k: int) -> int:
        return self.quickSort(nums, k, 0, len(nums)-1)
